client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -166,73 +167,159 @@ public Response performRequest(String method, String endpoint, Map<String, Strin
-            }
-            Response response = new Response(request.getRequestLine(), host, httpResponse);
-            int statusCode = response.getStatusLine().getStatusCode();
-            if (statusCode < 300 || (request.getMethod().equals(HttpHead.METHOD_NAME) && statusCode == 404) ) {
-                RequestLogger.logResponse(logger, request, host, httpResponse);
-                onResponse(host);
-                return response;
+            @Override
+            public void failed(Exception failure) {
+                try {
+                    RequestLogger.logFailedRequest(logger, request, host, failure);
+                    onFailure(host);
+                    retryIfPossible(failure, hosts, request);
+                } catch(Exception e) {
+                    listener.onDefinitiveFailure(e);
+                }
             }
-            RequestLogger.logResponse(logger, request, host, httpResponse);
-            String responseBody;
-            try {
-                if (response.getEntity() == null) {
-                    responseBody = null;
+
+            private void retryIfPossible(Exception exception, Iterator<HttpHost> hosts, HttpRequestBase request) {
+                if (hosts.hasNext()) {
+                    //in case we are retrying, check whether maxRetryTimeout has been reached
+                    long timeElapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+                    long timeout = maxRetryTimeoutMillis - timeElapsedMillis;
+                    if (timeout <= 0) {
+                        IOException retryTimeoutException = new IOException(
+                                "request retries exceeded max retry timeout [" + maxRetryTimeoutMillis + "]");
+                        listener.onDefinitiveFailure(retryTimeoutException);
+                    } else {
+                        listener.trackFailure(exception);
+                        request.reset();
+                        performRequest(startTime, hosts, request, responseConsumer, listener);
+                    }
                 } else {
-                    responseBody = EntityUtils.toString(response.getEntity());
+                    listener.onDefinitiveFailure(exception);
                 }
-            } finally {
-                response.close();
             }
-            lastSeenException = addSuppressedException(lastSeenException, new ResponseException(response, responseBody));
-            switch(statusCode) {
-                case 502:
-                case 503:
-                case 504:
-                    //mark host dead and retry against next one
-                    onFailure(host);
-                    break;
-                default:
-                    //mark host alive and don't retry, as the error should be a request problem
-                    onResponse(host);
-                    throw lastSeenException;
+
+            @Override
+            public void cancelled() {
```

Comment:
good point. I think it is ok because we don't expose the future and that is the only way to cancel it. This falls under the infamous category of "//this should never happen" kind of comments. Thoughts? :)